plugins/plant.cpp

```diff
@@ -0,0 +1,627 @@
+    int32_t age = -1; // Set plant to this age for grow/create; -1 for default
+
+    static struct_identity _identity;
+};
+static const struct_field_info plant_options_fields[] =
+{
+    { struct_field_info::PRIMITIVE, "create",    offsetof(plant_options, create),    &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "grow",      offsetof(plant_options, grow),      &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "del",       offsetof(plant_options, del),       &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "shrubs",    offsetof(plant_options, shrubs),    &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "saplings",  offsetof(plant_options, saplings),  &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "trees",     offsetof(plant_options, trees),     &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "dead",      offsetof(plant_options, dead),      &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "filter_ex", offsetof(plant_options, filter_ex), &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "zlevel",    offsetof(plant_options, zlevel),    &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "dry_run",   offsetof(plant_options, dry_run),   &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "plant_idx", offsetof(plant_options, plant_idx), &df::identity_traits<int32_t>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "age",       offsetof(plant_options, age),       &df::identity_traits<int32_t>::identity, 0, 0 },
+    { struct_field_info::END }
+};
+struct_identity plant_options::_identity(sizeof(plant_options), &df::allocator_fn<plant_options>, NULL, "plant_options", NULL, plant_options_fields);
+
+const int32_t sapling_to_tree_threshold = 120 * 28 * 12 * 3 - 1; // 3 years minus 1; let the game handle the actual growing-up
+
+command_result df_createplant(color_ostream &out, df::coord pos, const plant_options &options)
+{
+    auto col = Maps::getBlockColumn((pos.x / 48)*3, (pos.y / 48)*3);
+    if (!Maps::getTileBlock(pos) || !col)
+    {
+        out.printerr("No map block at pos!\n");
+        return CR_FAILURE;
+    }
+
+    auto tt = Maps::getTileType(pos);
+    if (!tt || tileShape(*tt) != tiletype_shape::FLOOR)
+    {
+        out.printerr("Plants can only be placed on floors!\n");
+        return CR_FAILURE;
+    }
+    else // Check tile mat and building occ
+    {
+        auto mat = tileMaterial(*tt);
+        if (mat != tiletype_material::SOIL &&
+            mat != tiletype_material::GRASS_DARK &&
+            mat != tiletype_material::GRASS_LIGHT)
+        {
+            out.printerr("Plants can only be placed on dirt or grass!\n");
```

Comment:
is this true for saplings? i thought they could grow on muddy natural stone floors

Reply:
It's just a limitation from the previous code. I could update the check like `regrass`.